File: server/routes/audit.js
/**
 * Audit Routes
 * Handles audit trail retrieval and filtering
 */

import express from 'express';
import { db } from '../database/db.js';

const router = express.Router();

// GET /api/audit - Get audit logs with optional filtering
router.get('/', (req, res) => {
  try {
    const { 
      userId, 
      action, 
      resourceType, 
      status, 
      startDate, 
      endDate, 
      limit = 100, 
      offset = 0 
    } = req.query;

    let query = 'SELECT * FROM audit_logs WHERE 1=1';
    const params = [];

    if (userId) {
      query += ' AND user_id = ?';
      params.push(userId);
    }
    if (action) {
      query += ' AND action LIKE ?';
      params.push(`%${action}%`);
    }
    if (resourceType) {
      query += ' AND resource_type = ?';
      params.push(resourceType);
    }
    if (status) {
      query += ' AND status = ?';
      params.push(status);
    }
    if (startDate) {
      query += ' AND created_at >= ?';
      params.push(startDate);
    }
    if (endDate) {
      query += ' AND created_at <= ?';
      params.push(endDate);
    }

    // Get total count
    const countQuery = query.replace('SELECT *', 'SELECT COUNT(*) as count');
    const totalCount = db.prepare(countQuery).get(...params).count;

    // Add ordering and pagination
    query += ' ORDER BY created_at DESC LIMIT ? OFFSET ?';
    params.push(parseInt(limit), parseInt(offset));

    const logs = db.prepare(query).all(...params);

    res.json({
      logs,
      pagination: {
        total: totalCount,
        limit: parseInt(limit),
        offset: parseInt(offset),
        hasMore: parseInt(offset) + logs.length < totalCount
      }
    });
  } catch (error) {
    console.error('Error fetching audit logs:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/audit/actions - Get distinct action types
router.get('/actions', (req, res) => {
  try {
    const actions = db.prepare('SELECT DISTINCT action FROM audit_logs ORDER BY action').all();
    res.json(actions.map(a => a.action));
  } catch (error) {
    console.error('Error fetching action types:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/audit/resources - Get distinct resource types
router.get('/resources', (req, res) => {
  try {
    const resources = db.prepare('SELECT DISTINCT resource_type FROM audit_logs WHERE resource_type IS NOT NULL ORDER BY resource_type').all();
    res.json(resources.map(r => r.resource_type));
  } catch (error) {
    console.error('Error fetching resource types:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/audit/stats - Get audit statistics
router.get('/stats', (req, res) => {
  try {
    const stats = {
      totalLogs: db.prepare('SELECT COUNT(*) as count FROM audit_logs').get().count,
      todayLogs: db.prepare(`
        SELECT COUNT(*) as count FROM audit_logs 
        WHERE DATE(created_at) = DATE('now')
      `).get().count,
      successCount: db.prepare(`
        SELECT COUNT(*) as count FROM audit_logs WHERE status = 'success'
      `).get().count,
      failureCount: db.prepare(`
        SELECT COUNT(*) as count FROM audit_logs WHERE status = 'failure'
      `).get().count,
      recentActions: db.prepare(`
        SELECT action, COUNT(*) as count 
        FROM audit_logs 
        GROUP BY action 
        ORDER BY count DESC 
        LIMIT 5
      `).all()
    };

    res.json(stats);
  } catch (error) {
    console.error('Error fetching audit stats:', error);
    res.status(500).json({ error: error.message });
  }
});

// GET /api/audit/:id - Get single audit log entry
router.get('/:id', (req, res) => {
  try {
    const log = db.prepare('SELECT * FROM audit_logs WHERE id = ?').get(req.params.id);
    if (!log) {
      return res.status(404).json({ error: 'Audit log not found' });
    }
    
    // Parse details JSON if present
    if (log.details) {
      try {
        log.details = JSON.parse(log.details);
      } catch (e) {
        // Keep as string if not valid JSON
      }
    }

    res.json(log);
  } catch (error) {
    console.error('Error fetching audit log:', error);
    res.status(500).json({ error: error.message });
  }
});

export { router as auditRoutes };

File: backend/app/routers/audit.py
"""
Audit Routes - Audit trail retrieval
"""
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from typing import Optional

from app.database import get_db, AuditLog
from app.utils.timezone import get_ph_now, format_ph_datetime

router = APIRouter()


@router.get("/")
async def list_audit_logs(
    userId: Optional[int] = None,
    action: Optional[str] = None,
    resourceType: Optional[str] = None,
    status: Optional[str] = None,
    limit: int = Query(default=100, le=500),
    offset: int = 0,
    db: AsyncSession = Depends(get_db)
):
    query = select(AuditLog)
    count_query = select(func.count(AuditLog.id))

    if userId:
        query = query.where(AuditLog.user_id == userId)
        count_query = count_query.where(AuditLog.user_id == userId)
    if action:
        query = query.where(AuditLog.action.contains(action))
        count_query = count_query.where(AuditLog.action.contains(action))
    if resourceType:
        query = query.where(AuditLog.resource_type == resourceType)
        count_query = count_query.where(AuditLog.resource_type == resourceType)
    if status:
        query = query.where(AuditLog.status == status)
        count_query = count_query.where(AuditLog.status == status)

    total_result = await db.execute(count_query)
    total = total_result.scalar()

    query = query.order_by(AuditLog.created_at.desc()).limit(limit).offset(offset)
    result = await db.execute(query)

    return {
        "logs": result.scalars().all(),
        "pagination": {
            "total": total,
            "limit": limit,
            "offset": offset,
            "hasMore": offset + limit < total
        }
    }


@router.get("/actions")
async def get_action_types(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(AuditLog.action).distinct().order_by(AuditLog.action))
    return [a for a in result.scalars().all()]


@router.get("/resources")
async def get_resource_types(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(AuditLog.resource_type).distinct()
        .where(AuditLog.resource_type.isnot(None))
        .order_by(AuditLog.resource_type)
    )
    return [r for r in result.scalars().all()]


@router.get("/stats")
async def get_audit_stats(db: AsyncSession = Depends(get_db)):
    total = await db.execute(select(func.count(AuditLog.id)))
    success = await db.execute(select(func.count(AuditLog.id)).where(AuditLog.status == "success"))
    failure = await db.execute(select(func.count(AuditLog.id)).where(AuditLog.status == "failure"))

    return {
        "totalLogs": total.scalar(),
        "successCount": success.scalar(),
        "failureCount": failure.scalar()
    }


@router.get("/{log_id}")
async def get_audit_log(log_id: int, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(AuditLog).where(AuditLog.id == log_id))
    log = result.scalar_one_or_none()
    if not log:
        raise HTTPException(status_code=404, detail="Audit log not found")
    return log


@router.get("/export/download")
async def export_audit_logs(
    format: str = Query(default="csv", description="Export format: csv, xlsx, json"),
    userId: Optional[int] = None,
    action: Optional[str] = None,
    resourceType: Optional[str] = None,
    status: Optional[str] = None,
    db: AsyncSession = Depends(get_db)
):
    """Export audit logs in CSV, XLSX, or JSON format."""
    import io
    import csv
    from datetime import datetime
    from fastapi.responses import StreamingResponse
    
    # Query audit logs
    query = select(AuditLog)
    if userId:
        query = query.where(AuditLog.user_id == userId)
    if action:
        query = query.where(AuditLog.action.contains(action))
    if resourceType:
        query = query.where(AuditLog.resource_type == resourceType)
    if status:
        query = query.where(AuditLog.status == status)
    
    result = await db.execute(query.order_by(AuditLog.created_at.desc()))
    logs = result.scalars().all()
    
    # Convert to list of dicts
    log_dicts = []
    for log in logs:
        log_dicts.append({
            "id": log.id,
            "timestamp": log.created_at.isoformat() if log.created_at else "",
            "user_id": log.user_id or "",
            "action": log.action or "",
            "resource_type": log.resource_type or "",
            "resource_id": log.resource_id or "",
            "status": log.status or "",
            "details": log.details or ""
        })
    
    timestamp_str = get_ph_now().strftime("%Y%m%d_%H%M%S")
    
    if format == "json":
        import json
        content = json.dumps(log_dicts, indent=2, default=str)
        return StreamingResponse(
            io.BytesIO(content.encode()),
            media_type="application/json",
            headers={"Content-Disposition": f"attachment; filename=audit_logs_{timestamp_str}.json"}
        )
    
    elif format == "xlsx":
        try:
            import openpyxl
            from openpyxl import Workbook
            
            wb = Workbook()
            ws = wb.active
            ws.title = "Audit Logs"
            
            # Headers
            headers = ["ID", "Timestamp", "User ID", "Action", "Resource Type", "Resource ID", "Status", "Details"]
            ws.append(headers)
            
            # Data
            for log in log_dicts:
                ws.append([log["id"], log["timestamp"], log["user_id"], log["action"], 
                          log["resource_type"], log["resource_id"], log["status"], log["details"]])
            
            output = io.BytesIO()
            wb.save(output)
            output.seek(0)
            
            return StreamingResponse(
                output,
                media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                headers={"Content-Disposition": f"attachment; filename=audit_logs_{timestamp_str}.xlsx"}
            )
        except ImportError:
            raise HTTPException(status_code=500, detail="openpyxl not installed for XLSX export")
    
    else:  # Default to CSV
        output = io.StringIO()
        writer = csv.DictWriter(output, fieldnames=["id", "timestamp", "user_id", "action", 
                                                     "resource_type", "resource_id", "status", "details"])
        writer.writeheader()
        writer.writerows(log_dicts)
        
        return StreamingResponse(
            io.BytesIO(output.getvalue().encode()),
            media_type="text/csv",
            headers={"Content-Disposition": f"attachment; filename=audit_logs_{timestamp_str}.csv"}
        )

